### ISSUE TITLE
关于文章中t-sne可视化的问题

### ISSUE THREAD

**Issue by wzh-moonlight** [h751410234/DATR#22]
大佬，求教这个文章中的那个t-sne可视化是怎么做到的。有代码可以提供或者参考吗？

**Comment by h751410234** [comment]
（1）修改网络推理代码，将需要可视化的特征保存为numpy格式（例如backbone的特征，或者object queries的特征）
（2）使用t-sne对特征进行可视化，可以参考下面的代码段：
import numpy as np
import os
from sklearn.manifold import TSNE
import matplotlib.pyplot as plt


def read_numpy_dir(dir_path,save_path):
    all_encoding_array = np.empty((0, 256))
    for i,base_name in enumerate(os.listdir(dir_path)):
        data_path = os.path.join(dir_path,base_name)
        encoding_array = np.load(data_path, allow_pickle=True)
        encoding_array = encoding_array.squeeze(0)
        all_encoding_array = np.append(all_encoding_array,encoding_array,axis=0)

    np.save(save_path, all_encoding_array)
    return all_encoding_array

def read_numpy_path(data_path):
    encoding_array = np.load(data_path, allow_pickle=True)
    return encoding_array


if __name__ == "__main__":

   # # #处理
   # #1.读取制备数据
   source_data_path = ''
   target_data_path = ''

   source_data = read_numpy_path(source_data_path)  #1878
   target_data = read_numpy_path(target_data_path)

    #2.合并data
   total_data_array = np.append(source_data,target_data,axis=0)


  # 3.T-SNE 画图
  # (1)初始化
   tsne = TSNE(n_components=2, n_iter=20000)
   #(2) 降维
   X_tsne_2d = tsne.fit_transform(total_data_array)
   print(X_tsne_2d.shape)

   #4.画图

   plt.figure(figsize=(14, 14))

   plt.scatter(X_tsne_2d[:18000, 0], X_tsne_2d[:18000, 1], color='brown', marker='.', label='source', s=150)
   plt.scatter(X_tsne_2d[18000:, 0], X_tsne_2d[18000:, 1], color='lightseagreen', marker='.', label='target', s=150)

   plt.legend(fontsize=16, markerscale=1, bbox_to_anchor=(1, 1))
   plt.xticks([])
   plt.yticks([])
   plt.savefig('语义特征t-SNE二维降维可视化_20.pdf', dpi=300)  # 保存图像
   plt.show()

**Comment by h751410234** [comment]
（1）修改网络推理代码，将需要可视化的特征保存为numpy格式（例如backbone的特征，或者object queries的特征）
（2）使用t-sne对特征进行可视化，可以参考下面的代码段：

![Image](https://github.com/user-attachments/assets/ff3d94a2-c906-404a-970e-5d2ad50d239e)

**Comment by wzh-moonlight** [comment]
成功了感谢感谢
